tests/infer/test_map.py

```diff
@@ -0,0 +1,27 @@
+import torch
+from torch.autograd import Variable
+
+import pyro
+import pyro.distributions as dist
+from pyro.infer.map import MAP
+from pyro.optim import Adam
+from pyro.util import ng_ones, ng_zeros
+from tests.common import assert_equal
+
+
+def test_map():
+
+    def model(data):
+        mu = pyro.param("mu", Variable(torch.zeros(1), requires_grad=True))
+        pyro.observe("mu_prior", dist.normal, mu, ng_zeros(1), ng_ones(1))
```

Comment:
@martinjankowiak feels that this `param`+`observe` pattern is awkward and that we should instead provide something like a `poutine.lower()` to automatically transform `sample` sites in this way. This will require some refactoring of `Poutine` and `apply_stack`, since current poutines are restricted to 1-1 site transformations, whereas `poutine.lower()` transforms one `sample` site into two sites. See #577 for details.